sdks/python/apache_beam/runners/interactive/user_pipeline_tracker_test.py

```
@@ -0,0 +1,181 @@
+
+    p = beam.Pipeline()
+
+    self.assertIsNone(ut.get_user_pipeline(p))
+
+  def test_no_parent_returns_none(self):
+    ut = UserPipelineTracker()
+
+    user = beam.Pipeline()
+    derived = beam.Pipeline()
+    orphan = beam.Pipeline()
+
+    ut.add_derived_pipeline(user, derived)
+
+    self.assertIsNone(ut.get_user_pipeline(orphan))
+
+  def test_get_user_pipeline_is_same(self):
+    ut = UserPipelineTracker()
+
+    p = beam.Pipeline()
+    ut.add_user_pipeline(p)
+
+    self.assertIs(ut.get_user_pipeline(p), p)
+
+  def test_can_add_derived(self):
+    ut = UserPipelineTracker()
+
+    user = beam.Pipeline()
+    derived = beam.Pipeline()
+
+    ut.add_derived_pipeline(user, derived)
+
+    self.assertIs(ut.get_user_pipeline(derived), user)
+
+  def test_can_add_multiple_derived(self):
+    """Tests that there can be many user pipelines with many derived
+    pipelines.
+    """
+    ut = UserPipelineTracker()
+
+    # Add the first set of user and derived pipelines.
+    user1 = beam.Pipeline()
+    derived11 = beam.Pipeline()
+    derived12 = beam.Pipeline()
+
+    ut.add_derived_pipeline(user1, derived11)
+    ut.add_derived_pipeline(user1, derived12)
+
+    # Add the second set of user and derived pipelines.
+    user2 = beam.Pipeline()
+    derived21 = beam.Pipeline()
+    derived22 = beam.Pipeline()
+
+    ut.add_derived_pipeline(user2, derived21)
+    ut.add_derived_pipeline(user2, derived22)
+
+    # Assert that the user pipelines are correct.
+    self.assertIs(ut.get_user_pipeline(derived11), user1)
+    self.assertIs(ut.get_user_pipeline(derived12), user1)
+    self.assertIs(ut.get_user_pipeline(derived21), user2)
+    self.assertIs(ut.get_user_pipeline(derived22), user2)
+
+  def test_cannot_have_multiple_parents(self):
+    ut = UserPipelineTracker()
+
+    user1 = beam.Pipeline()
+    user2 = beam.Pipeline()
+    derived = beam.Pipeline()
+
+    ut.add_derived_pipeline(user1, derived)
+    ut.add_derived_pipeline(user2, derived)
```

I wonder if this should throw an exception instead of silently failing?

---

or perhaps the return value should communicate this somehow

---

Added an exception